spec/models/team_spec.rb

```diff
@@ -147,6 +147,80 @@
     end
   end
 
+  describe 'limit of coaches' do
+    let(:team) { create :team }
+    let(:new_user) { create :user }
+    let(:coach_attributes) { [{ name: 'coach', team_id: team.id, user_id: new_user.id }] + 3.times.map { { name: 'coach', team_id: team.id, user_id: create(:user).id } } }
+    let(:new_coach_as_student) { { name: 'student', team_id: team.id, user_id: new_user.id } }
+    let(:fifth_new_coach) { { name: 'coach', team_id: team.id, user_id: create(:user).id } }
+    let(:remove_coach) { { name: 'coach', team_id: team.id, user_id: create(:user).id, _destroy: true } }
+
+    context 'when team has no coaches yet' do
+      it 'allows to add 2 new coaches' do
+        expect {
+          team.update roles_attributes: coach_attributes
+        }.to change { team.members.count }.by 4
+      end
+
+      it 'ignores coaches marked for destruction' do
+        roles_attributes = coach_attributes
+        roles_attributes << remove_coach
+        expect {
+          team.update roles_attributes: roles_attributes
+        }.to change { team.members.count }.by 4
+      end
+
+      it 'does not allow to add more than 5 new coaches' do
+        roles_attributes = coach_attributes
+        roles_attributes << fifth_new_coach
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:roles].first).to eql 'there cannot be more than 4 coaches on a team.'
+      end
+
+      it 'does not allow the same coach to fill up both spots' do
+        roles_attributes = 2.times.map { coach_attributes.first }
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:base].first).to eql "#{new_user.name} can't have more than one role in this team!"
+      end
+
+      it 'does not allow the same coach to add themselves as a student' do
+        roles_attributes = [coach_attributes.first, new_coach_as_student]
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:base].first).to eql "#{new_user.name} can't have more than one role in this team!"
+      end
+    end
+    context 'when team has coaches' do
+      let!(:existing_coaches) { 4.times.map { create :coach_role, team: team, user: create(:user) } }
```

Comment:
The `4.times.map` works, of course, but it looks a bit hacky to me. What about 

```ruby
let!(:existing_coaches) { create_list :coach_role, 4, team: team }  # `user` is implied in the factory
```

Reply:
This worked great, thank you!